### DISCUSSION TITLE
Writing a python function and exposing it to the OpenFOAM layer

### DISCUSSION THREAD

**Discussion by uttamcadambi07** [mdolab/dafoam#403] [General]
Hi,

In the DAFoam Workshop 2022, it is very clear on how to add a functionality in the OpenFOAM layer and expose it to the Python layer by wrapping it using cython. 

However, I would like to know how to do the reverse. I have a Python code that does some advanced operations on the fields that are obtained from OpenFOAM. I want to be able to pass these fields to my Python code, do the operations and send back data to the OpenFOAM layer to use it for further computation.

Can you please provide some direction on where I can start looking? I am guessing if I could create a dummy OpenFOAM function and declare it in pyDASolvers.py script in a more detailed fashion, will it work?

**Comment by friedenhe** [comment]
A lot of existing functions are already doing what you want. Take a look at the [calcdRdXvTPsi](https://github.com/mdolab/dafoam/blob/b3f73e247456e7526789e0dc0458dc134c9474da/dafoam/mphys/mphys_dafoam.py#L657) function. It passes a seed vec psiVec (computed in the Python layer) to the OpenFOAM layer, and the OpenFOAM layer will compute [dR/dXv]^T * psi and return it as prodVec.

**Reply by uttamcadambi07** [reply]
Perfect. I checked it out. Seems like wVec is the OF vector fields converted into a state vector. And xVec is the mesh points. Another question - what is the ordering for the state vector? I printed it and noticed that the ordering seems a little confusing. For example, the streamwise velocity appears every 11th index for the first N indices and then the frequency shifts to every 10th index and ends with something else.  I want the streamwise velocity that corresponds to a given cell centroid and I want them as a Python array so that I can manipulate it,

**Reply by friedenhe** [reply]
Check the [adjoint paper](https://www.researchgate.net/publication/333518918_Effective_Adjoint_Approaches_for_Computational_Fluid_Dynamics)'s appendix on the state order.

**Reply by uttamcadambi07** [reply]
Thanks a lot! I have managed to pass the velocity field alone as a petsc vector! So I collect the velocity fields from all processors onto the master processor and then dump it into one petsc vector which I then use convertMPIVec2SeqVec to convert to a sequential vector and then an array in the PYDAFOAM class in pyDAFoam.py. I need to still check if it is efficient to do this. Will keep the forum posted.
